feat(captor): integrating Kivebpf in Koney

San7o opened the pull request:
## Description

This PR integrates Kivebpf (tag v1.0.0-pre) in Koney. Kivebpf is a free and open source eBPF-powered file access monitoring Kubernetes operator. Kivebpf improves Koney by providing a light and fast captor alternative to Tetragon, specifically designed for Koney.

Key improvements include:

- fast alert time (less than 1 second) compared to Tetragon (sometimes more than 10s)
- simpler logic compared to Tetragon, making Kivebpf easy to integrate, maintain and extend
- inode-based monitoring
- support for regexes in container names
- Kivebpf will evolve to meet Koney's use cases

Current limitations for Kivebpf:

- limited support for container runtimes and architectures (currently, only `containerd` and x86_64 are supported, broader support will come over time)
- `volumeMount` decoy strategy is not supported
- the current working directory and the arguments of the process that accessed the file may not be reported if the process lived for a short amount of time

## Testing

New end-to-end tests were added for Kivebpf. Tested on a Kind cluster with Kubernetes 1.33, Linux 6.14 x86_64 (Ubuntu).

Reporting alerts with the sink to Dynatrace was not tested since, from my understanding, it is still in development.

## Checklist

- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Code builds without warnings
- [x] Tests pass locally
- [x] Documentation updated

## Additional Notes

The `containerSelector` was changed to use regexes compatible with go's regexp package, providing a more complete and expressive regex matching usage. This PR closes #22 .

Dependencies have been upgraded (go 1.24, Kubernetes 1.33).

Changes are based on commit d7c4ecc3c7e6a4a0735700df25286f585fa93a46  (branch [feat/TR-4934-dynatrace-ingest](https://github.com/dynatrace-oss/koney/tree/feat/TR-4934-dynatrace-ingest)).

blu3r4y reviewed (changes requested):
Thanks @San7o for this AMAZING pull request, very clean code, very attentive, very consistent with the rest of the case, lots of tiny improvements. I love your coding discipline! 🥳 

I tried my best to give decent review. I did not yet run the tests myself because I ran into issues trying to deploy Kive.

San7o commented on alert-forwarder/forwarder/kive.py:
```diff
@@ -0,0 +1,46 @@
+def process_kive_alert(kiveAlert: json) -> KoneyAlert:
+    koneyAlert = KoneyAlert(
+        timestamp=kiveAlert["timestamp"],
+        deception_policy_name=kiveAlert["kive-policy-name"],
```
The name of the `KivePolicy` is set to be the name of the `TracingPolicy` that generated it, since there is a relationship 1 to 1 between the two.

San7o replied:
A better option would be to add additional metadata to the generated `KivePolicy`, for example:

```yaml
apiVersion: kivebpf.san7o.github.io/v2alpha1
kind: KivePolicy
metadata:
  labels:
    app.kubernetes.io/name: kivebpf
  finalizers:
    - kivepolicy.kivebpf.san7o.github.io/finalizer
  name: kive-sample-policy
  namespace: kivebpf-system
spec:
  traps:
  - path: /secret.txt
    matchAny:
    - matchLabels:
        security-level: high
    metadata:
      security: high
```

That last `metadata` field would be propagated to the alert in a `custom-metadata` json field. This could also be used to support #27

San7o commented on docs/DEVELOPER_GUIDE.md:
```diff
@@ -8,7 +8,7 @@ This document describes how to build, test, and deploy the Koney operator.
 - `kubectl` version v1.11.3+.
 - Access to a Kubernetes v1.11.3+ cluster.
-- [Tetragon](https://tetragon.io/) v1.1.0+ installed in the cluster, if you also want to monitor traps.
+- [Tetragon](https://tetragon.io/) v1.1.0+ or [Kive](https://github.com/San7o/kivebpf) installed in the cluster, if you also want to monitor traps.
```
`cert-manager` is needed for resource versioning (support for v1, v2alpha1 and subsequent versions). The principle is similar to validation webhooks, in the sense that the resource will first be passed to the resource versioning conversion hook and then processed. It is required to use TLS, and `cert-manager` or another certificate provider is needed.

blu3r4y replied:
Hmm ... if this is really necessary, okay. I am mainly asking because this introduces a new mandatory dependency for all users of Kive, and then, also for all users of Koney. Also, I was surprised because I installed a lot of operators already and never needed cert-manager as a hard requirement before.

I just looked into Kyverno, who also use validation webhooks. Interestingly enough the bundle their own cert generation and renewal package:

* https://github.com/kyverno/kyverno/blob/4903223e6211d89a163a1242118f2f21dcf048be/pkg/tls/keypair.go#L17
* https://kyverno.io/docs/installation/customization/#certificate-management

San7o replied:
I thought that self-signed certificates were usually a bad idea. Even operator SDK warns about it:

```
// TODO(user): TLSOpts is used to allow configuring the TLS config used for the server. If certificates are
// not provided, self-signed certificates will be generated by default. This option is not recommended for
// production environments as self-signed certificates do not offer the same level of trust and security
// as certificates issued by a trusted Certificate Authority (CA). The primary risk is potentially allowing
// unauthorized access to sensitive metrics data. Consider replacing with CertDir, CertName, and KeyName
// to provide certificates, ensuring the server communicates using trusted and secure certificates.
```
https://github.com/dynatrace-oss/koney/blob/main/cmd/main.go#L107

So I went with `cert-manager` which seems to solve the security problem but adds a new dependency.
It surprises me hearing that Kubernetes applications do not use a trusted CA in their clusters, I thought
this was a common dependency. I also thought that, when Koney will use webhooks or have more versions
then it would depend on `cert-manager` anyway for said security reasons.

If it is the case that `cert-manager` is an unwanted dependency, the operator will need to figure out how
to manage its certificates and secrets correctly, which I am skeptical of. I can look into how to do this
and see how doable it is.

San7o commented on internal/controller/matching/matching.go:
```diff
@@ -320,11 +320,13 @@ func selectContainers(resource client.Object, containerSelector string) ([]strin
 
+	compiledRegex, err := regexp.Compile(containerSelector)
+	if err != nil {
+		return []string{}, fmt.Errorf("compiling regex: %w", err)
```
Sure, this is great.

San7o commented on internal/controller/traps/filesystoken/deployment.go:
```diff
@@ -413,6 +421,33 @@ func (r *FilesystemHoneytokenReconciler) deployCaptorWithTetragon(ctx context.Co
+		return err
+	}
+
+	if err := r.Client.Patch(ctx, tracingPolicy, client.Apply, client.ForceOwnership, client.FieldOwner("koney-controller")); err != nil {
```
Correct, we do not need to check if it exists and it can be removed from `deployCaptorWithTetragon`

San7o commented on internal/controller/traps/filesystoken/utils.go:
```diff
@@ -264,3 +268,86 @@ func generateTetragonTracingPolicy(deceptionPolicy *v1alpha1.DeceptionPolicy, tr
+			Labels: map[string]string{
+				constants.LabelKeyDeceptionPolicyRef: deceptionPolicy.Name,
+			},
+			Finalizers: []string{kivev1.KivePolicyFinalizerName},
```
Yes, it makes sense that the responsibility should be of Kive. I will remove them from this function.

San7o commented on alert-forwarder/forwarder/main.py:
I upgraded `golangci` since the tool complains that he needs a newer or equal version than the version of the go compiler.

The upgraded version is now smarter and when `unparam` is set in `.golangci.yml` (which is the case) he recognizes when a function should return an error but never actually does so. To stop the `pre-commit` from failing I had to change the following function from this:

```
func generateTetragonTracingPolicy(deceptionPolicy *v1alpha1.DeceptionPolicy, trap v1alpha1.Trap, tracingPolicyName string) (*ciliumiov1alpha1.TracingPolicy, error) {
``` 

to this:

```
func generateTetragonTracingPolicy(deceptionPolicy *v1alpha1.DeceptionPolicy, trap v1alpha1.Trap, tracingPolicyName string) *ciliumiov1alpha1.TracingPolicy {
```

I did the same for `generateKivePolicy`, and I changed all the places where these functions were called.

San7o commented on internal/controller/matching/matching_test.go:
This is indeed an unfortunate issue. I think prepending the
string with "blob:" or "regex:" or nothing is overall the best
solution, as users can decide how powerful / complex matching
can be. By default, it should be the simplest solution that does
not use any regex.

San7o commented:
@blu3r4y Since the original review, I have:

- fixed all obvious mistakes / typos from the code and documentation https://github.com/dynatrace-oss/koney/pull/28/commits/727f01fd08160a3805ab5b4d42120f92527a367f
- added support for volume mounts https://github.com/San7o/kivebpf/pull/7
- implemented custom user metadata in `KivePolicy`s that will be present in their `KiveAlert`s https://github.com/San7o/kivebpf/pull/8
- added support for multiple expression for matching container names, namely `regex:`, `glob:` and exact comparison https://github.com/San7o/kivebpf/pull/9
- made `Kive` more robust, fast, and fault tolerant, with better error handling https://github.com/San7o/kivebpf/commit/27ded6f5ce4d84920a0716528457d78c23843ad6, https://github.com/San7o/kivebpf/commit/cff3b205ffbf9339f4084efd7c910ae7d2c5b310, https://github.com/San7o/kivebpf/commit/136c7938f18b2daa2dc453a6929ce4deaa884718

You can review the changes again and run the end to end tests with the operator deployed. Current `Kivebpf` commit is 57f317730da5a51d4a85895820c5e5446fef043c, which will be released as v1.0.0-pre2

blu3r4y commented on internal/controller/traps/filesystoken/utils_test.go:
```diff
@@ -27,9 +27,9 @@ import (
-		"namewithwildcard.*",
-		"namewithwildcard?",
-		".*",
+		"glob:namewithwildcard.*",
```
[ISSUE] I think the dot is a left-over that is not needed for the glob syntax.

```suggestion
		"glob:namewithwildcard*",
```

blu3r4y commented on internal/controller/traps/filesystoken/utils_test.go:
```diff
@@ -93,7 +92,7 @@ var _ = Describe("generateTetragonTracingPolicy", func() {
 				// Check the container selector
 				for _, resourceFilter := range trap.MatchResources.Any {
-					if resourceFilter.ResourceDescription.ContainerSelector == "" || resourceFilter.ResourceDescription.ContainerSelector == ".*" {
+					if resourceFilter.ResourceDescription.ContainerSelector == "" || resourceFilter.ResourceDescription.ContainerSelector == "regex:.*" {
```
[NIT] How about using `ContainerSelectorSelectsAll` from `matchutils.go` here?

blu3r4y commented:
> [DISCUSSION] Regarding the issue of not consistently / incorrectly reporting the current working directory and the arguments of the process that accessed a file, I would propose to disable the reporting of these features completely for the release versions of Kive until a solution is found. The reason is that I would prefer Kive v1.0.0 to have a set of fully-working, reliable features and a stable API. Having unreliable `KiveAlert`s is not ideal: once a solution is implemented, they will be reported, but for the meantime the fields should simply be empty strings.
> 
> What do you think? @blu3r4y

@San7o I agree, I think it's acceptable to disable this right now. Referencing #26 so that we mention this in the technical support matrix eventually.

blu3r4y commented on internal/controller/utils/matching.go:
```diff
@@ -0,0 +1,51 @@
+	"strings"
+)
+
+func MatchContainerName(regex string, containerName string) (bool, error) {
```
[CLARITY] It can be other things than `regex`
```suggestion
func MatchContainerName(pattern string, containerName string) (bool, error) {
```

blu3r4y commented on README.md:
```diff
@@ -120,7 +120,13 @@ The `any` field is a list and holds one or more `resources` objects, which conta
-- `containerSelector`: selects the container(s) in the matched pods or deployments where the trap is deployed. It supports the same pattern syntax as [`regexp`](https://github.com/google/re2/wiki/Syntax). The default value is empty, and the trap is deployed in all containers in the matched pods.
+- `containerSelector`: selects the container(s) in the matched pods or deployments where the trap is deployed.
+  - if this field is prepended by `regex:`, the rest of the string will represent a regular expression matched with go [regexp](https://golang.org/s/re2syntax) library.
+  - if the field is prepended by "glob:", then this is a filesystem-style regex, as described in go [filepath.Match](https://pkg.go.dev/path/filepath#Match) library
```
[FORMATTING, CLARITY] I think the proper name is "globbing" (https://en.wikipedia.org/wiki/Glob_(programming))
```suggestion
  - if the field is prepended by `glob:`, then this is a glob pattern, as described in go [filepath.Match](https://pkg.go.dev/path/filepath#Match) library
```

San7o commented on alert-forwarder/forwarder/sink.py:
```diff
@@ -52,6 +52,17 @@ def read_alert_sinks() -> list[AlertSink]:
+        return alert_sinks
+    except:
+        if logger.level <= logging.ERROR:
+            console.print(K8S_SINK_READ_ERROR, style="bold red")
```
"Python continues to be the most loved programming language in 2025" - All surveys

blu3r4y commented:
Hi @San7o, I rebased the PR again because `main` changed notably in the meantime because we are preparing Koney to be installed via a Helm chart. It's working already, it's just not released yet because I also want to embed Kive into it next. The only noteworthy changes that I made during the rebase are the removal of `constants.KivePolicyNamespace` and `constants.KiveWebhookUrl` because they are now resolved dynamically to support scenarios where users installed Koney in a namespace other than `koney-system`. Also, our full test suite is now running in GitHub actions and I see that all of them passed.

🥳 

**Thanks again for the PR !! I am going to merge this PR within the next few days, if you don't have any objections?**

---

Further, would you mind tagging a new release e.g. `v1.0.0-pre3` of Kive's current main branch, and pushing that to the GitHub container registry (GHCR)? The recent changes with the self-signed certificates are only in `dev` right now, which is why we need to install Kive like this for now as a workaround:

```
kubectl apply -f https://raw.githubusercontent.com/San7o/kivebpf/refs/heads/main/dist/install-remote.yaml
kubectl set image -n kivebpf-system daemonset kivebpf-manager "*=ghcr.io/san7o/kivebpf:dev"
```

Also, I see that kivebpf is also pushed to Docker Hub currently (https://hub.docker.com/r/giovann103/kivebpf) and that version is also currently referenced in `install-remote.yaml` but the Docker Hub versions wasn't pushed in a long time. I think it would be better to either (a) push to both GHCR and Docker Hub and keep them in-sync, or (b) only push to GHCR and reference `ghcr.io/san7o/kivebpf` in all manifests.
